File: problemas/estruturas-e-bibliotecas/1062/1062.js
var input = require('fs').readFileSync('/dev/stdin', 'utf8');
var lines = input.split('\n');

class Pilha {
    constructor() {
        this.pilha = [];
    }

    push(valor) {
        this.pilha.push(valor);
    }

    pop() {
        this.pilha.pop();
    }

    top() {
        return this.pilha[this.pilha.length - 1];
    }

    size() {
        return this.pilha.length;
    }

    empty() {
        return this.pilha.length === 0;
    }
}

while(lines.length){
    let N = parseInt(lines.shift());

    if(!N)  break;

    let numeros = lines.shift().trim().split(' ').map((x) => parseInt(x));
    while(numeros.length > 1 || numeros[0] !== 0){
        let A = new Pilha();
        let estacao = new Pilha();
        let B = new Pilha();

        for(let i = 0; i < numeros.length; ++i){
            A.push(numeros[i]);
            B.push(i + 1);
        }

        while(!A.empty() || !estacao.empty() || !B.empty()){
            if(!A.empty() && !B.empty() && A.top() === B.top()){
                A.pop();
                B.pop();
            }else if(!estacao.empty() && !B.empty() && estacao.top() === B.top()){
                estacao.pop();
                B.pop();
            }else if(!A.empty()){
                estacao.push(A.top());
                A.pop();
            }else{
                break;
            }
        }

        if(A.empty() && estacao.empty() && B.empty()){
            console.log("Yes");
        }else{
            console.log("No");
        }

        numeros = lines.shift().trim().split(' ').map((x) => parseInt(x));
    }
    console.log("");
}

File: problemas/estruturas-e-bibliotecas/1062/1062.cpp
#include <iostream>
#include <stack>

using namespace std;

int main(){
    int N, x;
    stack<int> A, estacao, B;

    while(cin >> N){
        if(!N)  break;

        while(cin >> x){
            if(!x){
                cout << endl;
                break;
            }

            while(!A.empty())       A.pop();
            while(!estacao.empty()) estacao.pop();
            while(!B.empty())       B.pop();

            A.push(x);
            B.push(1);
            for(int i = 2; i <= N; ++i){
                cin >> x;

                A.push(x);
                B.push(i);
            }

            while(!A.empty() || !estacao.empty() || !B.empty()){
                if(!A.empty() && !B.empty() && A.top() == B.top()){
                    A.pop();
                    B.pop();
                }else if(!estacao.empty() && !B.empty() && estacao.top() == B.top()){
                    estacao.pop();
                    B.pop();
                }else if(!A.empty()){
                    estacao.push(A.top());
                    A.pop();
                }else{
                    break;
                }
            }

            if(A.empty() && estacao.empty() && B.empty()){
                cout << "Yes" << endl;
            }else{
                cout << "No" << endl;
            }
        }
    }

    return 0;
}
